### PR TITLE
Add: Add a standalone Plugin for `ChangedOID` check

### PR DESCRIPTION
**Pull request by pascalholthaus**
**What**:  changed_oid plugin as standalone

**Why**:
DEVOPS-214

**How**:

**Checklist**:

- [ ] Tests
- [ ] Conventional commit message
- [ ] Documentation

### CODE REVIEW

File: troubadix/standalone_plugins/changed_oid.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            )
+            yield nasl_file
+
+
+def main(args) -> int:
+    try:
+        git_base = git("rev-parse", "--show-toplevel")
+        os.chdir(git_base.stdout.rstrip("\n"))
+    except subprocess.SubprocessError:
+        print("In your current working directory is no git repository")
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:137]:
```suggestion
        print("Your current working directory doesn't belong to a git repository")
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        help=(
+            "List of files to diff."
+            "If empty use all files added or modifyed in the"
+            " commit range"
+        ),
+    )
+    return parser.parse_args(args=args)
+
+
+def git(*args) -> subprocess.CompletedProcess:
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:69]:
```suggestion
def git(*args) -> str:
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    return subprocess.run(
+        ["git"] + list(args), capture_output=True, encoding="utf-8", check=True
+    )
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:72]:
```suggestion
    ).stdout
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+        print(f"Check file {nasl_file}")
+        text = git(
+            "-c",
+            "color.status=false",
+            "--no-pager",
+            "diff",
+            args.commit_range,
+            nasl_file,
+        ).stdout
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:105]:
```suggestion
        )
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                f"OID_NEW: {oid_added.group('oid')} "
+                f"OID OLD: {oid_removed.group('oid')}"
+            )
+            yield nasl_file
+
+
+def main(args) -> int:
+    try:
+        git_base = git("rev-parse", "--show-toplevel")
+        os.chdir(git_base.stdout.rstrip("\n"))
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:135]:
```suggestion
        os.chdir(git_base.rstrip("\n"))
```

**Review by bjoernricks — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 6]
```diff
@@ -78,6 +78,7 @@ version-module-file = "troubadix/__version__.py"
 
 [tool.poetry.scripts]
 troubadix = 'troubadix.troubadix:main'
+troubadix-change-oid = 'troubadix.standalone_plugins.changed_oid:main'
```
Comment (Dexus) [inline pyproject.toml:81]:
```suggestion
troubadix-changed-oid = 'troubadix.standalone_plugins.changed_oid:main'
```

**Review by Dexus — CHANGES_REQUESTED** [review]

File: tests/standalone_plugins/test_changed_oid.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        test_file.write_text(
+            'script_oid("1.3.6.1.4.1.25623.1.0.100313");\ntest'
+        )
+    else:
+        test_file.write_text('script_oid("2.3.6.1.4.1.25623.1.0.100313");')
+    git("add", "-u")
+    git("commit", "-m", "test2")
+
+
+class TestChangeOid(unittest.TestCase):
```
Comment (Dexus) [inline tests/standalone_plugins/test_changed_oid.py:63]:
```suggestion
class TestChangedOid(unittest.TestCase):
```

**Review by Dexus — CHANGES_REQUESTED** [review]

**Review by Dexus — APPROVED** [review]

**Comment by n-thumann** [comment]
I also ran this script against a branch of of the vts repo. I then changed the OID of a VT that has been modified in the previous commit, added and commited the changes.
When I now run troubadix against `HEAD^`, it will report the change, but not if run against `HEAD^^` (or `HEAD^^..HEAD`)
```
➜  vts git:(notus2nasl_for_debian_2013) troubadix-changed-oid -c HEAD^ 
Check file scripts/2013/debian/deb_2834_1.nasl
Change OID IN scripts/2013/debian/deb_2834_1.nasl OID_NEW: 1.3.6.1.4.1.25623.1.0.70283499999 OID OLD: 1.3.6.1.4.1.25623.1.0.702834
```
```
➜  vts git:(notus2nasl_for_debian_2013) troubadix-changed-oid -c HEAD^^
Check file scripts/2013/debian/deb_2389_1.nasl
[...]
Check file scripts/2013/debian/deb_2834_1.nasl
```

Is this expected?

File: troubadix/standalone_plugins/changed_oid.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            print(
+                f"Change OID IN {nasl_file} "
+                f"OID_NEW: {oid_added.group('oid')} "
+                f"OID OLD: {oid_removed.group('oid')}"
+            )
+            rcode = True
+    return rcode
+
+
+def main(args) -> int:
```
[the remark this answers is not in the document]
Comment (pascalholthaus) [reply troubadix/standalone_plugins/changed_oid.py]: Thanks there was an error with the pip installation

**Comment by pascalholthaus** [comment]
I think I got it, the git filter was probably set a little too sharply.

**Comment by y0urself** [comment]
> I think I got it, the git filter was probably set a little too sharply.

@n-thumann can you do us a favor and test this again?

**Comment by n-thumann** [comment]
> @n-thumann can you do us a favor and test this again?

I'm already in touch with @pascalholthaus, since the issue still persists for me :)

**Comment by pascalholthaus** [comment]
If you add a file in a first commit and then change it with a second commit. It may be that git diff only displays the changes made to the commit before adding the new file as a new file.

File: troubadix/standalone_plugins/changed_oid.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    - script_oid("1.2.3");
+
+    This is only allowed in rare cases (e.g. a single VT was split into
+    two VTs).
+    """
+
+    if not args.files:
+        args.files += [
+            Path(_file)
+            for _file in git(
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:88]:
```suggestion
            Path(f)
            for f in git(
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        return 1
+
+    if check_oid(parse_args(args)):
+        return 2
+
+    return 0
+
+
+if __name__ == "__main__":
+    sys.exit(main(sys.argv[1:]))
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:154]:
```suggestion
    sys.exit(main())
```

File: troubadix/standalone_plugins/changed_oid.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+                f"OID_NEW: {oid_added.group('oid')} "
+                f"OID OLD: {oid_removed.group('oid')}"
+            )
+            rcode = True
+    return rcode
+
+
+def main(args=None) -> int:
+    if not args:
+        args = sys.argv[1:]
```
Comment (bjoernricks) [inline troubadix/standalone_plugins/changed_oid.py:136]:
```suggestion
def main() -> int:
    args = sys.argv[1:]
```

**Review by bjoernricks — APPROVED** [review]

**Pull request merged**